### PR TITLE
Convert Tokenizers By Default

### PR DESCRIPTION
**Pull request by apaniukov**
# What does this PR do?

Make tokenizer conversion [the default behaviour](https://github.com/huggingface/optimum-intel/pull/513#discussion_r1507086074). 
- Add `--disable-convert-tokenizer` option.
- Add a warning for the old `--convert-tokenizer` option.
- Add detailed advanced check for tokenizers version compatibility and detailed instructions for troubleshooting.

Covered scenarios for different OpenVINO Tokenizers \ OpenVINO combinations:
| Tokenizers \ OpenVINO   | Release PyPI | Pre-release Simple PyPI | -nightly PyPI | Archive |
| ----------------------- | :----------: | :---------------------: | :-----------: | :-----: |
| Release PyPI            |      +       |            +            |       +       |    +    |
| Pre-release Simple PyPI |      +       |            +            |       +       |    +    |
| Build                   |              |                         |               |    +    |

The tokenizer conversion adds a couple of seconds and the resulting files are a couple of megabytes. For opt-125m (no pytorch weights download), the conversion without tokenizer took 13.340s compared to 13.969s with tokenizer conversion, which is less than 5%. If we take weights downloading time into account, the percentage is even lower.
The resulting folder size increases from 482 MB to 484 MB.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

### CODE REVIEW

**Review by eaidova — APPROVED** [review]

**Comment by eaidova** [comment]
it seems useless until we need to install openvino tokenizers as extra explicitly, please consider including it in openvino extra

**Comment by slyalin** [comment]
> it seems useless until we need to install openvino tokenizers as extra explicitly, please consider including it in openvino extra

@apaniukov, please add openvino_tokenizer in openvino extra in this PR.

File: optimum/intel/utils/import_utils.py
Comment (helena-intel) [inline optimum/intel/utils/import_utils.py:98]: This will break when users install OpenVINO with other methods than `pip install openvino`.

Comment (apaniukov) [reply optimum/intel/utils/import_utils.py]: Update checks. Add a table for covered scenarios.

File: optimum/intel/utils/import_utils.py
Comment (helena-intel) [inline optimum/intel/utils/import_utils.py:247]: It would be good to explain which branch to pick, because building the source version only fixes a version conflict if the version the user has installed happens to be the same as the then current source version.

File: optimum/intel/utils/import_utils.py
Comment (helena-intel) [inline optimum/intel/utils/import_utils.py:241]: image
Since the error message said "first 3 numbers" I expected these slightly different nightly versions would work, but it doesn't. So then it makes more sense to say that the versions have to be exactly the same.

I get this error even when I use an archive of March 27 https://storage.openvinotoolkit.org/repositories/openvino/packages/nightly/2024.1.0-14876-fc5da9d9359 and openvino tokenizers  2024.1.0.0.dev20240327 
image

**Review by helena-intel — COMMENTED** [review]
I created comments about the error messages as I was testing this, but there are quite a few things that can go wrong or cause confusion. IMO it is a better user experience to just show one line like "OpenVINO tokenizer model could not be exported. This does not impact model export. See [documentation url] for more info" and then explain there how to make sure all versions are in sync.

File: optimum/intel/utils/import_utils.py
Comment (apaniukov) [reply optimum/intel/utils/import_utils.py]: That is why it is just `info`, not a warning. But I agree, replace the message with a link to installation instructions.

File: optimum/intel/utils/import_utils.py
Comment (apaniukov) [reply optimum/intel/utils/import_utils.py]: Yes, but there is the next line with the command to install the pre-release version, which is identical to nightly but won't cause problems with two openvino versions in one environment.

File: optimum/intel/utils/import_utils.py
Comment (apaniukov) [reply optimum/intel/utils/import_utils.py:230]: This is because of the different Linux distros. Add this message if Linux platform is detected:
> The PyPI version of OpenVINO Tokenizers is built on CentOS and may not be compatible with other Linux distributions; rebuild OpenVINO Tokenizers from source.

File: optimum/intel/utils/import_utils.py
Comment (apaniukov) [reply optimum/intel/utils/import_utils.py:236]: This is not the case, the compatibility is defined by the OpenVINO version, not OpenVINO Tokenizers version. The only issue is if the archive OpenVINO version is old and OpenVINO Tokenizers use a newer API.

I can add a message about release branches in the OpenVINO Tokenizers repo instead.

File: tests/openvino/test_exporters_cli.py
Comment (echarlaix) [inline tests/openvino/test_exporters_cli.py:122]: we shoudl remove :
```diff
- @unittest.skipIf(not is_openvino_tokenizers_available(), reason="OpenVINO Tokenizers not available")
```

File: optimum/intel/utils/import_utils.py
Comment (echarlaix) [inline optimum/intel/utils/import_utils.py:173]: where is this used ?

File: optimum/intel/utils/import_utils.py
Comment (echarlaix) [inline optimum/intel/utils/import_utils.py:197]: would prefer
```suggestion
    if importlib.util.find_spec("openvino_tokenizers") is None:
```

File: optimum/intel/utils/import_utils.py
Comment (echarlaix) [inline optimum/intel/utils/import_utils.py:214]: this should be done outside in my opinion (check if nightly)

File: optimum/intel/utils/import_utils.py
Comment (echarlaix) [inline optimum/intel/utils/import_utils.py:262]: this warnings should only be displayed during the export step so should be moved somewhere else

File: optimum/intel/utils/import_utils.py
Comment (apaniukov) [reply optimum/intel/utils/import_utils.py:236]: Update readme in [this PR](https://github.com/openvinotoolkit/openvino_tokenizers/pull/104)

File: tests/openvino/test_exporters_cli.py
Comment (apaniukov) [reply tests/openvino/test_exporters_cli.py:121]: Removed and add logic checks for ["Test openvino-nightly" stage](https://github.com/huggingface/optimum-intel/blob/a48e0ca3177075235c2a67ee770e07bd751e49dd/.github/workflows/test_openvino.yml#L39), where the libraries are not compatible.

File: optimum/intel/utils/import_utils.py
Comment (apaniukov) [reply optimum/intel/utils/import_utils.py]: Fixed

File: optimum/intel/utils/import_utils.py
Comment (apaniukov) [reply optimum/intel/utils/import_utils.py]: Do you mean in the separate function? It doesn't matter for other components, so I check it here.

File: optimum/intel/utils/import_utils.py
Comment (apaniukov) [reply optimum/intel/utils/import_utils.py]: Moved [here](https://github.com/huggingface/optimum-intel/blob/eb055941103db6a2101df93741afd0c6be34dabd/optimum/exporters/openvino/__main__.py#L347).

File: optimum/exporters/openvino/__main__.py
Comment (echarlaix) [inline optimum/exporters/openvino/__main__.py:323]: The tokenizer export results in additional files being included which could be confusing for users, can this be moved to to an `openvino_tokenizers` directory ? 

```suggestion
    if convert_tokenizer and is_openvino_tokenizers_available():
         output = Path(output) / "openvino_tokenizers"
```

File: tests/openvino/test_exporters_cli.py
Comment (echarlaix) [inline tests/openvino/test_exporters_cli.py:134]: would prefer to have it removed to make sure this is always tested (we should always have a compatible version of `openvino` / `openvino-tokenizer` when testing)
```suggestion
```

File: tests/openvino/test_exporters_cli.py
Comment (echarlaix) [inline tests/openvino/test_exporters_cli.py:123]: we should also check whether no error was raised here

Comment (echarlaix) [reply tests/openvino/test_exporters_cli.py:123]: With
```bash
optimum-cli export openvino --model hf-internal-testing/tiny-random-t5 --task text2text-generation ov_model
```

I'm getting the following error : 
```
OpenVINO Tokenizer export for T5TokenizerFast is not supported. Exception: [Errno 2] No such file or directory: '/tmp/tmprj8zsg44/spiece.model'
```

Comment (echarlaix) [reply tests/openvino/test_exporters_cli.py:123]: Would like to have this fixed before making the tokenizer export default / merging this PR

File: optimum/exporters/openvino/__main__.py
Comment (apaniukov) [reply optimum/exporters/openvino/__main__.py:323]: This is similar to the original tokenizer, so we replicate an existing pattern here.

File: tests/openvino/test_exporters_cli.py
Comment (apaniukov) [reply tests/openvino/test_exporters_cli.py:123]: It is not a bug, we don't support the Unigram model from `tokenizer.json` yet, only from sentencepiece model file. If you use a [repository](https://huggingface.co/google-t5/t5-base/tree/main) that has such a file, the tokenizer will be converted.

I added one more check for the vocab file here: https://github.com/openvinotoolkit/openvino_tokenizers/pull/116
This will transform
```shell
OpenVINO Tokenizer export for T5TokenizerFast is not supported. Exception: [Errno 2] No such file or directory: '/tmp/tmprj8zsg44/spiece.model'
```
into:
```shell
OpenVINO Tokenizer export for T5TokenizerFast is not supported. Exception: Cannot convert tokenizer of this type without `.model` file.
```

The issue is that the original tokenizer has info about `.model` file in the tokenizer object, but it does not have it in practice:
image

I would argue that this is an edge case of a test repo and most tokenizers with info about the `.model` file have it, so it should not block the merge.

File: optimum/exporters/openvino/__main__.py
Comment (echarlaix) [reply optimum/exporters/openvino/__main__.py:323]: I'm thinking about something similar than what's done for [SD models](https://huggingface.co/stabilityai/stable-diffusion-xl-base-1.0/tree/main/tokenizer) is this something that sounds reasonable to you? I'm fine with adding it in a following PR if you prefer not including it in this one

Comment (apaniukov) [reply optimum/exporters/openvino/__main__.py:323]: Yes, I can do that in a separate PR.

File: tests/openvino/test_exporters_cli.py
Comment (apaniukov) [reply tests/openvino/test_exporters_cli.py:123]: It doesn't throw an error in this case, just prints an error message from a tokenizer conversion process. The model is converted fine:
```bash
>> optimum-cli export openvino --model hf-internal-testing/tiny-random-t5 --task text2text-generation openvino_model
INFO:nncf:NNCF initialized successfully. Supported frameworks detected: torch, onnx, openvino
Framework not specified. Using pt to export the model.
The task `text2text-generation` was manually specified, and past key values will not be reused in the decoding. if needed, please pass `--task text2text-generation-with-past` to export using the past key values.
Using the export variant default. Available variants are:
    - default: The default ONNX variant.
Using framework PyTorch: 2.1.2+cu121
Overriding 1 configuration item(s)
        - use_cache -> False
Using framework PyTorch: 2.1.2+cu121
Overriding 1 configuration item(s)
        - use_cache -> False
/home/apaniuko/python/optimum-intel/venv/lib/python3.10/site-packages/transformers/modeling_utils.py:908: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  if causal_mask.shape[1] < attention_mask.shape[1]:
OpenVINO Tokenizer export for T5TokenizerFast is not supported. Exception: [Errno 2] No such file or directory: '/tmp/tmpyww_b5i_/spiece.model'

>> ls openvino_model/
config.json             openvino_decoder_model.bin  openvino_encoder_model.bin  special_tokens_map.json  tokenizer.json
generation_config.json  openvino_decoder_model.xml  openvino_encoder_model.xml  tokenizer_config.json
```

The confusion comes from the fact that this is the last message from the conversion process. Maybe adding the "The model converted successfully" message will be beneficial. This message can be added in a separate PR as well.

Comment (echarlaix) [reply tests/openvino/test_exporters_cli.py:123]: Can we either disable this warning or disable export for unsupported cases ?

Comment (apaniukov) [reply tests/openvino/test_exporters_cli.py:123]: This warning is for any exception that appeared during conversion. After seeing this message, the user can create an issue for a particular model/tokenizer support.
I also prefer to tell the user that the tokenizer is not converted, rather than silently not including it without warning, so the lack of a (de)tokenizer model won't be a surprise.

Comment (echarlaix) [reply tests/openvino/test_exporters_cli.py:123]: I'd agree if the export of the tokenizer was an explicit choice of the user (current `--convert-tokenizer`) but as this PR makes it default I think having such warning can be an issue as it makes it look like the export failed

Comment (apaniukov) [reply tests/openvino/test_exporters_cli.py:123]: I changed the log level to debug, the message won't be visible by default.

Comment (echarlaix) [reply tests/openvino/test_exporters_cli.py:123]: thanks @apaniukov

**Comment by echarlaix** [comment]
Couple of tests are failing, will merged once fixed, let me know if you need any help @apaniukov

**Comment by echarlaix** [comment]
Now that https://github.com/huggingface/optimum-intel/pull/618 is merged, the tokenizer won't be exported by default when hybrid quantization is applied for [SD models](https://github.com/huggingface/optimum-intel/blob/4651ac2c4a05e047c86c66a3347b39cb6384cd0b/optimum/commands/export/openvino.py#L214-L247) I'm fine with merging it now and for us to add this in a following PR (with https://github.com/huggingface/optimum-intel/pull/580#discussion_r1562231029) what do you think @apaniukov ?

**Comment by apaniukov** [comment]
> Now that #618 is merged, the tokenizer won't be exported by default when hybrid quantization is applied for [SD models](https://github.com/huggingface/optimum-intel/blob/4651ac2c4a05e047c86c66a3347b39cb6384cd0b/optimum/commands/export/openvino.py#L214-L247) I'm fine with merging it now and for us to add this in a following PR (with [#580 (comment)](https://github.com/huggingface/optimum-intel/pull/580#discussion_r1562231029)) what do you think @apaniukov ?

No problem, let's do that and I will create a new PR with the fixes we discussed here.

**Pull request merged**
